flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkSessionJobController.java

```diff
@@ -0,0 +1,240 @@
+    }
+
+    @Override
+    public List<EventSource> prepareEventSources(
+            EventSourceContext<FlinkSessionJob> eventSourceContext) {
+        Preconditions.checkNotNull(controllerConfig, "Controller config cannot be null");
+        Set<String> effectiveNamespaces = controllerConfig.getEffectiveNamespaces();
+        if (effectiveNamespaces.isEmpty()) {
+            return List.of(createFlinkDepInformerEventSource(ALL_NAMESPACE));
+        } else {
+            return effectiveNamespaces.stream()
+                    .map(this::createFlinkDepInformerEventSource)
+                    .collect(Collectors.toList());
+        }
+    }
+
+    private InformerEventSource<FlinkDeployment, FlinkSessionJob> createFlinkDepInformerEventSource(
+            String name) {
+        return new InformerEventSource<>(
+                kubernetesClient.resources(FlinkDeployment.class).runnableInformer(0),
+                primaryResourceRetriever(),
+                sessionJob ->
+                        new ResourceID(
+                                sessionJob.getSpec().getClusterId(),
+                                sessionJob.getMetadata().getNamespace()),
+                false) {
+            @Override
+            public String name() {
+                return name;
+            }
+        };
+    }
+
+    /**
+     * Mapping the {@link FlinkDeployment} session cluster to {@link FlinkSessionJob}. It leverages
+     * the informer indexer.
+     *
+     * @return The {@link PrimaryResourcesRetriever}.
+     */
+    private PrimaryResourcesRetriever<FlinkDeployment> primaryResourceRetriever() {
+        return flinkDeployment -> {
+            var namespace = flinkDeployment.getMetadata().getNamespace();
+            var informer =
+                    controllerConfig.getEffectiveNamespaces().isEmpty()
+                            ? informers.get(ALL_NAMESPACE)
+                            : informers.get(namespace);
+
+            var sessionJobs =
+                    informer.getIndexer()
+                            .byIndex(CLUSTER_ID_INDEX, flinkDeployment.getMetadata().getName());
+            var resourceIDs = new HashSet<ResourceID>();
+            for (FlinkSessionJob sessionJob : sessionJobs) {
+                resourceIDs.add(
+                        new ResourceID(
+                                sessionJob.getMetadata().getName(),
+                                sessionJob.getMetadata().getNamespace()));
+            }
+            LOG.debug(
+                    "Find the target resource {} for {} ",
+                    resourceIDs,
+                    flinkDeployment.getMetadata().getNamespace());
+            return resourceIDs;
+        };
+    }
+
+    /**
+     * Create informers for session job to build indexer for cluster to session job relations.
+     *
+     * @return The different namespace's index informer.
+     */
+    private Map<String, SharedIndexInformer<FlinkSessionJob>> createInformers() {
```

Comment:
To clarify here: we maintain an extra informer index outside the controller internal informer for the `FlinkSessionJob`. Because it do not support to build indexer on top of it before preparing the event sources. I have create an [issue](https://github.com/java-operator-sdk/java-operator-sdk/issues/1114) to discuss with java-operator-sdk community about, It seems can be supported upstream, After that we can save this indexer builder.